### PR TITLE
Add more type hints to HTTP client.

### PR DESCRIPTION
**Pull request by clokep**
Following up on #8806, this:

* Adds more type hints to the HTTP client code.
* Combines some duplicated code.

### CODE REVIEW

**Comment by richvdh** [comment]
@clokep did you mean #8806?

**Comment by clokep** [comment]
> @clokep did you mean #8806?

I did in fact! 🤦

File: synapse/http/client.py [hunk 1]
```diff
@@ -732,35 +764,27 @@ def connectionLost(self, reason):
...
+    if args is None:
+        return b""
 
-def encode_urlencode_arg(arg):
-    if isinstance(arg, str):
-        return arg.encode("utf-8")
-    elif isinstance(arg, list):
-        return [encode_urlencode_arg(i) for i in arg]
-    else:
-        return arg
```
Comment (richvdh) [inline synapse/http/client.py:755]: Can we be absolutely certain that nothing was relying on being able to pass in `bytes`es here?

File: synapse/http/matrixfederationclient.py [hunk 2]
```diff
@@ -28,26 +28,27 @@
...
     HttpResponseException,
     RequestSendFailed,
-    SynapseError,
 )
 from synapse.http import QuieterFileBodyProducer
-from synapse.http.client import BlacklistingAgentWrapper, IPBlacklistingResolver
+from synapse.http.client import (
+    BlacklistingAgentWrapper,
+    IPBlacklistingResolver,
+    _readBodyToFile,
```
Comment (richvdh) [inline synapse/http/matrixfederationclient.py:49]: it'd be nice to rename this if we're going to import it across modules.

**Comment by richvdh** [comment]
lgtm other than those two thoughts

File: synapse/http/client.py [hunk 3]
```diff
@@ -732,35 +764,27 @@ def connectionLost(self, reason):
...
+        encoded_args[k] = [v.encode("utf8") for v in vs]
 
+    query_str = urllib.parse.urlencode(encoded_args, True)
 
-def _print_ex(e):
-    if hasattr(e, "reasons") and e.reasons:
-        for ex in e.reasons:
-            _print_ex(ex)
-    else:
-        logger.exception(e)
```
Comment (clokep) [inline synapse/http/client.py:763]: I couldn't find any callers to this, not sure when it became unused.

File: synapse/http/matrixfederationclient.py (hunk 2 above)
Comment (clokep) [reply synapse/http/matrixfederationclient.py:49]: Good call. I thought of that and then just failed to do it. 😢

File: synapse/http/client.py (hunk 1 above)
Comment (clokep) [reply synapse/http/client.py:755]: This was only used by `post_urlencoded_get_json`, which is called in two places:

* [`IdentityHandler.ask_id_server_for_third_party_invite`](https://github.com/matrix-org/synapse/blob/1c262431f9bf768d106bf79a568479fa5a0784a1/synapse/handlers/identity.py#L858-L860)
    * Following this up the chain of callers it seems the keys and values are always strings.
* [`RecaptchaAuthChecker.check_auth`](https://github.com/matrix-org/synapse/blob/c9c0ad5e204f309f2686dbe250382e481e0f82c2/synapse/handlers/ui_auth/checkers.py#L108)
    * Following the callers it seems like the data is strings, in particular it seems the [response is a string and not bytes](https://github.com/matrix-org/synapse/blob/514a240aedd972e6c8b707a7685c3d283f8618c9/synapse/rest/client/v2_alpha/auth.py#L129-L134)

Would appreciate a double checking of this though.

**Comment by clokep** [comment]
@richvdh I think I've answered your questions! Thanks for taking a look.

File: synapse/http/client.py (hunk 1 above)
Comment (richvdh) [reply synapse/http/client.py:755]: yeah, seems fair to me :+1:

File: synapse/http/client.py [hunk 4]
```diff
@@ -730,7 +730,7 @@ def _timeout_to_request_timed_out_error(f: Failure):
     return f
 
 
-class _ReadBodyToFileProtocol(protocol.Protocol):
+class readBodyToFileProtocol(protocol.Protocol):
```
Comment (richvdh) [inline synapse/http/client.py:733]: class names start with capital letters.

**Review by richvdh — APPROVED** [review]
lgtm otherwise

File: synapse/http/client.py (hunk 4 above)
Comment (clokep) [reply synapse/http/client.py:733]: Ugh, I didn't actually mean to rename this. There's no reason for it to be public. Shoddy work, sorry about that.

**Pull request merged**